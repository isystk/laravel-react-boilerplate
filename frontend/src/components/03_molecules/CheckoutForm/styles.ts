import { css, keyframes } from '@emotion/css'

export const checkoutForm = css`
  label > div {
    background-color: white;
  }

  label.selected div {
    background-color: black;
    color: white;
  }
`
