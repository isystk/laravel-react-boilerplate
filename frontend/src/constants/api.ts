import { ENDPOINT_URL } from './index'

/**
 * BFF（バックエンドフォーフロントエンド）用の URL を作成する
 */

/** API のエンドポイント */
export const Api = {
  /* Public */
  Product: [ENDPOINT_URL, '/product'].join(''),
  Payment: [ENDPOINT_URL, '/payment'].join(''),
  CancelRequest: [ENDPOINT_URL, '/cancel-request'].join(''),
  CancelConfirm: [ENDPOINT_URL, '/cancel-confirm'].join(''),
  Cancel: [ENDPOINT_URL, '/cancel'].join(''),

  /* Auth */
  Login: [ENDPOINT_URL, '/login'].join(''),
  LoginCheck: [ENDPOINT_URL, '/login-check'].join(''),
  Logout: [ENDPOINT_URL, '/logout'].join(''),

  /* Private */
  AdminSubscriber: [ENDPOINT_URL, '/admin/subscriber'].join(''),
  AdminSubscriberTrend: [ENDPOINT_URL, '/admin/subscriberTrend'].join(''),
  AdminCancel: [ENDPOINT_URL, '/admin/cancel'].join(''),
}
