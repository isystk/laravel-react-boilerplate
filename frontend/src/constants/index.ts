//  ----------------------------------------------------------------------------
//  System const values
//  ----------------------------------------------------------------------------

/** システム アプリ名 */
const APP_NAME = process.env.APP_NAME

/** システム アプリ説明 */
const APP_DESCRIPTION = process.env.APP_DESCRIPTION

/** システム デフォルトロケール */
const LOCALE = process.env.NEXT_PUBLIC_LOCALE as 'ja' | 'en'

/** システム モード */
const APP_MODE = process.env.NODE_ENV

/** システム APP URL */
const APP_URL = process.env.NEXT_PUBLIC_APP_URL

/** Dateオブジェクト 文字列表現デフォルトフォーマット */
const DATE_FORMAT = 'yyyy-MM-dd HH:mm:ss'

/** エンドポイント URL */
const ENDPOINT_URL = process.env.NEXT_PUBLIC_ENDPOINT_URL

/** Stripe 公開鍵 */
const STRIPE_KEY = process.env.NEXT_PUBLIC_STRIPE_KEY

/** Google Analytics トラッキングID */
const GA_TRAKING_ID = process.env.NEXT_PUBLIC_GA_TRAKING_ID

type LOCALSTORAGE = {
  User
}
export type LOCALSTORAGE_KEYS = keyof LOCALSTORAGE

export {
  APP_NAME,
  APP_DESCRIPTION,
  LOCALE,
  APP_MODE,
  APP_URL,
  DATE_FORMAT,
  ENDPOINT_URL,
  STRIPE_KEY,
  GA_TRAKING_ID,
}
