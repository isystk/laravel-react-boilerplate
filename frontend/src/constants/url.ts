import { APP_URL } from './index'

/**
 * フロントエンド用の URL を作成する
 */

export const Url = {
  Top: [APP_URL, '/'].join(''),
  Product: [APP_URL, '/product/prod_NpvV9ohJIlgElI'].join(''),
  Payment: [APP_URL, '/product/prod_NpvV9ohJIlgElI/payment'].join(''),
  Cancel: [APP_URL, '/product/prod_NpvV9ohJIlgElI/cancel'].join(''),

  /* 管理画面 */
  AdminLogin: [APP_URL, '/admin/login'].join(''),
  AdminHome: [APP_URL, '/admin/home'].join(''),
  AdminProduct: [APP_URL, '/admin/product'].join(''),
  AdminSubscriber: [APP_URL, '/admin/subscriber'].join(''),
}
