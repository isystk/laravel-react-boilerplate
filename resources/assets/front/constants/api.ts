import Env from '@/constants/env';

const getBffUrl = (path: string): string => {
  return [Env.ENDPOINT_URL, path].join('');
};

/** API のエンドポイント */
export const Api = {
  /** ログイン状態チェック */
  LOGIN_CHECK: getBffUrl('/loginCheck'),
  /** ログイン */
  LOGIN: getBffUrl('/authenticate'),
  /** ログアウト */
  LOGOUT: getBffUrl('/logout'),
  /** 共通定数 */
  CONST: getBffUrl('/const'),

  /** お気に入りデータ取得 */
  LIKE: getBffUrl('/like'),
  /** お気に入り追加 */
  LIKE_STORE: getBffUrl('/like/store'),
  /** お気に入り削除 */
  LIKE_DESTROY: getBffUrl('/like/destroy'),

  /** お問い合わせ登録 */
  CONTACT_STORE: getBffUrl('/contact/store'),

  /** 商品一覧データ取得 */
  STOCK: getBffUrl('/stock'),
  /** マイカートデータ取得 */
  MYCART: getBffUrl('/mycart'),
  /** カートに商品を追加する */
  MYCART_ADD: getBffUrl('/mycart/add'),
  /** カートから商品を削除する */
  MYCART_DELETE: getBffUrl('/mycart/delete'),
  /** Stripe用のペイメント作成 */
  MYCART_PAYMENT: getBffUrl('/mycart/payment'),
  /** 決算処理後の後処理 */
  MYCART_CHECKOUT: getBffUrl('/mycart/checkout'),
};
