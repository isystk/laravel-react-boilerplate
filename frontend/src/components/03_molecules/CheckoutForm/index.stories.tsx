import { Meta, Story } from '@storybook/react'
import React from 'react'
import CheckoutForm, { CheckoutFormProps } from './index'
import { Elements } from '@stripe/react-stripe-js'
import { STRIPE_KEY } from '@/constants'
import { loadStripe } from '@stripe/stripe-js'
const stripePromise = loadStripe(STRIPE_KEY)

export default {
  title: '03_molecules/CheckoutForm',
  component: CheckoutForm,
} as Meta

const Template: Story = () => {
  const props: CheckoutFormProps = {
    product: {
      plans: [
        {
          id: 1,
          amount: 999,
          currency: 'jpy',
        },
      ],
    },
  }
  return (
    <Elements stripe={stripePromise}>
      <CheckoutForm {...props} />
    </Elements>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
