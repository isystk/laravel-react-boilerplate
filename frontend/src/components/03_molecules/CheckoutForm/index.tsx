import React, { FC, useState } from 'react'
import { ContainerProps } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import Loading from '@/components/01_atoms/Loading'
import Input from '@/components/01_atoms/Input'
import { CardElement, useElements, useStripe } from '@stripe/react-stripe-js'
import { useForm, type SubmitHandler } from 'react-hook-form'
import axios, { AxiosError } from '@/utils/axios'
import { PaymentMethodResult } from '@stripe/stripe-js'
import { Api } from '@/constants/api'
import * as stripe from 'stripe'
import { useI18n } from '@/components/i18n'

const CARD_ELEMENT_OPTIONS = {
  hidePostalCode: true, // 郵便番号を非表示
  style: {
    base: {
      fontFamily: 'Sans Serif',
      fontStyle: 'normal',
      fontWeight: '400',
      fontSize: '16px',
      letterSpacing: '1px',
      color: '#333333',
      '::placeholder': {
        color: '#a9a9a9',
      },
    },
  },
}

type FormInputs = {
  name: string
  email: string
  planId: string
}

export type ProductData = {
  plans: stripe.Plan[]
} & stripe.Product

/** CheckoutFormProps Props */
export type CheckoutFormProps = {
  product: ProductData
  userKey?: string
}
/** Presenter Props */
export type PresenterProps = CheckoutFormProps & {
  t
  onsubmit
  product
  isComplete
  loading
  cardErrorMsg
  errorMsg
  register
  handleSubmit
  errors
  stripe
}

/** Presenter Component */
const CheckoutFormPresenter: FC<PresenterProps> = ({
  t,
  onsubmit,
  product,
  isComplete,
  loading,
  cardErrorMsg,
  errorMsg,
  register,
  handleSubmit,
  errors,
  validate,
  stripe,
  planId,
  setPlanId,
  ...props
}) => (
  <>
    <div>
      {!isComplete ? (
        <form onSubmit={handleSubmit(onsubmit)} className={styles.checkoutForm}>
          <div className="flex flex-wrap items-center md:mb-8">
            {product.plans.map(({ id, amount, currency, interval }, idx) => {
              const locale = navigator.language
              const amountFmt = amount
                ? new Intl.NumberFormat(locale, {
                    style: 'currency',
                    currency,
                  }).format(amount)
                : ''
              return (
                <label
                  key={id}
                  className={`${
                    planId === id ? 'selected' : ''
                  } w-full md:w-1/2`}
                  onClick={() => setPlanId(id)}
                >
                  <div className="border text-center rounded-lg md:mr-4 py-12 h-72 mb-8">
                    <div className="mb-12">
                      {interval === 'month'
                        ? t('monthly amount')
                        : interval === 'year'
                        ? t('yearly amount')
                        : 'その他'}
                    </div>
                    <div className="font-bold text-5xl mb-12">{amountFmt}</div>
                    <input
                      type="radio"
                      value={id}
                      {...register('planId', validate['planId'])}
                    />
                  </div>
                </label>
              )
            })}
            {errors.planId && (
              <span className="pt-4 text-red-500">{errors.planId.message}</span>
            )}
          </div>
          <div className="flex flex-col mb-4">
            <Input
              {...{
                placeholder: t('your name'),
                type: 'text',
                name: 'name',
                register,
                validate,
                errors,
              }}
            />
          </div>
          <div className="flex flex-col mb-4">
            <Input
              {...{
                placeholder: t('Email address'),
                type: 'email',
                name: 'email',
                register,
                validate,
                errors,
              }}
            />
          </div>
          <div className="p-4 bg-gray-200 rounded-md">
            <CardElement options={CARD_ELEMENT_OPTIONS} />
            {cardErrorMsg && (
              <span className="pt-4 text-red-500">{cardErrorMsg}</span>
            )}
          </div>
          <div>
            <button
              className="bg-blue-500 rounded-md text-white border-none rounded-5 px-20 py-4 text-18 cursor-pointer mt-10"
              type="submit"
              disabled={!stripe}
            >
              {t('Buy Now')}
            </button>
          </div>
          {errorMsg && <p className="pt-4 text-red-500">{errorMsg}</p>}
        </form>
      ) : (
        <p className="text-18 text-left mb-20">
          {t('The purchase has been completed.')}
        </p>
      )}
      <Loading loading={loading} />
    </div>
  </>
)

/** Container Component */
const CheckoutFormContainer: React.FC<
  ContainerProps<CheckoutFormProps, PresenterProps>
> = ({ presenter, product, userKey, ...props }) => {
  const { t } = useI18n('Common')

  const [isComplete, setIsComplete] = useState(false)
  const [loading, setLoading] = useState(false)
  const [cardErrorMsg, setCardErrorMsg] = useState('')
  const [errorMsg, setErrorMsg] = useState('')
  const stripe = useStripe()
  const elements = useElements()
  const [planId, setPlanId] = useState(product.plans[0].id)

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<FormInputs>()

  const validate = {
    planId: {
      required: t('Plan Required'),
    },
    name: {
      required: t('Please enter your name'),
      max: {
        value: 30,
        message: t('Please enter your name in 30 characters or less'),
      },
    },
    email: {
      required: t('Please enter your email address'),
      pattern: {
        value: /[\w\-\\._]+@[\w\-\\._]+\.[A-Za-z]+/,
        message: t('Email address is correct'),
      },
    },
  }

  // フォーム送信ボタンを押された時の処理
  const onsubmit: SubmitHandler<FormInputs> = async ({
    planId,
    name,
    email,
  }) => {
    try {
      // create a payment method
      const payment: PaymentMethodResult | undefined =
        await stripe?.createPaymentMethod({
          type: 'card',
          card: elements?.getElement(CardElement),
          billing_details: {
            email,
          },
        })
      if (!payment) {
        setCardErrorMsg(t('Please enter your card information') as string)
        return
      }
      if (payment.error) {
        setCardErrorMsg(
          payment.error.message ?? (t('Invalid card information') as string)
        )
        return
      }
      setLoading(true)

      // 決済処理をする
      const { data } = await axios.post(Api.Payment, {
        userKey: userKey ?? email,
        paymentMethod: payment.paymentMethod?.id,
        name,
        email,
        planId,
      })
      reset()

      setIsComplete(true)
    } catch (e: unknown) {
      console.log(e)
      if (e instanceof AxiosError) {
        const { response } = e
        setErrorMsg(t(response?.data?.message) as string)
      } else if (e instanceof Error) {
        setErrorMsg(e.message)
      }
    } finally {
      setLoading(false)
    }
  }

  return presenter({
    t,
    onsubmit,
    product,
    isComplete,
    loading,
    cardErrorMsg,
    errorMsg,
    register,
    handleSubmit,
    errors,
    validate,
    stripe,
    planId,
    setPlanId,
    ...props,
  })
}

export default connect<CheckoutFormProps, PresenterProps>(
  'CheckoutForm',
  CheckoutFormPresenter,
  CheckoutFormContainer
)
