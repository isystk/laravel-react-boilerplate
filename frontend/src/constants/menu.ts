import { Url } from './url'

export type MenuItem = {
  label: string
  href: string
  target?: '_blank'
}

/**
 * メニュー を管理する
 */
export const frontMenuItems: MenuItem[] = [
  {
    label: 'Home',
    href: Url.Top,
  },
  {
    label: 'Rates',
    href: Url.Payment,
  },
  {
    label: 'Contact Us',
    href: 'https://blog.isystk.com/contact/',
    target: '_blank',
  },
]

export const frontFooterMenuItems: MenuItem[] = [
  {
    label: 'Company',
    href: 'https://blog.isystk.com/company/',
    target: '_blank',
  },
  {
    label: 'Privacy Policy',
    href: 'https://blog.isystk.com/privacy-poricy/',
    target: '_blank',
  },
  {
    label: 'Contact Us',
    href: 'https://blog.isystk.com/contact/',
    target: '_blank',
  },
  {
    label: 'Twitter',
    href: 'https://twitter.com/ise0615',
    target: '_blank',
  },
]

export const adminMenuItems: MenuItem[] = [
  {
    label: 'Product List',
    href: Url.AdminProduct,
  },
  {
    label: 'Subscribe List',
    href: Url.AdminSubscriber,
  },
]
