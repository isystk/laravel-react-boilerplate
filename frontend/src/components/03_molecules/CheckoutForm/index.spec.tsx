import React from 'react'
import renderer from 'react-test-renderer'
import CheckoutForm, { CheckoutFormProps } from './index'
import { Elements } from '@stripe/react-stripe-js'
import { STRIPE_KEY } from '@/constants'
import { loadStripe } from '@stripe/stripe-js'
const stripePromise = loadStripe(STRIPE_KEY)

describe('CheckoutForm', () => {
  it('Match Snapshot', () => {
    const props: CheckoutFormProps = {
      product: {
        plans: [
          {
            id: 1,
            amount: 999,
            currency: 'jpy',
          },
        ],
      },
    }
    const component = renderer.create(
      <Elements stripe={stripePromise}>
        <CheckoutForm {...props} />
      </Elements>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
