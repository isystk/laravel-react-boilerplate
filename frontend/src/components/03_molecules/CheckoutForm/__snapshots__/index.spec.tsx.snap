// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm Match Snapshot 1`] = `
<div>
  <form
    className="css-11os301"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-wrap items-center md:mb-8"
    >
      <label
        className="selected w-full md:w-1/2"
        onClick={[Function]}
      >
        <div
          className="border text-center rounded-lg md:mr-4 py-12 h-72 mb-8"
        >
          <div
            className="mb-12"
          >
            その他
          </div>
          <div
            className="font-bold text-5xl mb-12"
          >
            ¥999
          </div>
          <input
            name="planId"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value={1}
          />
        </div>
      </label>
    </div>
    <div
      className="flex flex-col mb-4"
    >
      <input
        className="css-0 p-3 w-full bg-gray-200 rounded-md"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="your name"
        type="text"
      />
    </div>
    <div
      className="flex flex-col mb-4"
    >
      <input
        className="css-0 p-3 w-full bg-gray-200 rounded-md"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email address"
        type="email"
      />
    </div>
    <div
      className="p-4 bg-gray-200 rounded-md"
    >
      <div />
    </div>
    <div>
      <button
        className="bg-blue-500 rounded-md text-white border-none rounded-5 px-20 py-4 text-18 cursor-pointer mt-10"
        disabled={true}
        type="submit"
      >
        Buy Now
      </button>
    </div>
  </form>
  <div
    className="css-0 fixed top-0 left-0 w-full h-full z-30 hidden"
  >
    <div
      className="absolute bottom-0 right-0 p-8 flex justify-center"
    >
      <div
        className="animate-spin h-10 w-10 border-4 border-gray-500 rounded-full border-t-transparent"
      />
    </div>
  </div>
</div>
`;
